src/blockrelay/graphene.cpp

```diff
@@ -337,38 +336,39 @@ bool CGrapheneBlock::CheckBlockHeader(const CBlockHeader &block, CValidationStat
  */
 bool CGrapheneBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, std::string strCommand, unsigned nHops)
 {
-    int nSizeGrapheneBlock = vRecv.size();
-    CInv inv(MSG_BLOCK, uint256());
+    // Deserialize grapheneblock and store a block to reconstruct
+    CGrapheneBlock tmp(NegotiateGrapheneVersion(pfrom), NegotiateFastFilterSupport(pfrom));
+    vRecv >> tmp;
+    auto pblock = thinrelay.SetBlockToReconstruct(pfrom, tmp.header.GetHash());
+    pblock->grapheneblock = std::make_shared<CGrapheneBlock>(std::forward<CGrapheneBlock>(tmp));
 
-    CGrapheneBlock grapheneBlock(NegotiateGrapheneVersion(pfrom), NegotiateFastFilterSupport(pfrom));
-    vRecv >> grapheneBlock;
+    std::shared_ptr<CGrapheneBlock> grapheneBlock = pblock->grapheneblock;
```

Comment:
This might be picky, but if you do:
globalVar = malloc()
localVar = globalVar
you risk another thread modifying globalVar in between those 2 lines.  Maybe that can't happen here anyway due to locks, BUT if you do:
localVar = malloc()
globalVar = localVar

the code shows that it can't happen without knowing anything about the rest of program state.   So I think doing it that way is better.

Reply:
I modified SetBlockToReconstruct() so that once it's set for a particular node and hash it won't be re-intitialized by any susequent attempts.   At any rate , the way it was we were holding the smart pointer already to graphenblock so if another thread modified it in pblock we would still have our own separate smart point to our data so it wouldn't impact the processing of HandleMessage...but if there was a re-request, then you'd end up using the new pointer to grapheneblock and potentially messing up your reconstruction, but it wouldn't crash...but for now it's all moot because we would never download two grapheneblocks for the same blockhash anyway, and furthermore we can't currently download two grapheneblocks at the same time...Regardless you bring up good points that need to be considered and I think the changes in SetBlockToReconstruct() take care of our current needs.